File: app/src/main/java/com/app/sid/funwithflags/view/countrylist/CountryListInteractor.java
package com.app.sid.funwithflags.view.countrylist;

import com.app.sid.funwithflags.datasets.remote.Country;

import java.util.List;

import rx.Observable;
import rx.functions.Action1;
import rx.functions.Func1;

/**
 * Created by Sidhanshu Udawat on 19-Jan-18.
 */

public class CountryListInteractor implements CountryListMvp.Interactor {

    private final CountryListMvp.RemoteDataSource mRemoteDataSource;
    private final CountryListMvp.LocalDataSource mLocalDataSource;

    public CountryListInteractor(CountryListMvp.LocalDataSource localDataSource, CountryListMvp.RemoteDataSource remoteDataSource) {
        mLocalDataSource = localDataSource;
        mRemoteDataSource = remoteDataSource;
    }

    @Override
    public Observable<List<Country>> getCountryList() {
        if (mLocalDataSource.isLocalDataPresent()) {
            return mLocalDataSource.getCountryList();
        } else {
            return mRemoteDataSource.getCountryList()
                    .flatMap(countries -> Observable.from(countries).doOnNext(mLocalDataSource::saveCountry))
                    .toList();
        }
    }
}

File: app/src/main/java/com/app/sid/funwithflags/view/countrylist/CountryListInteractor.kt
package com.app.sid.funwithflags.view.countrylist

import com.app.sid.funwithflags.datasets.remote.Country
import com.app.sid.funwithflags.view.countrylist.CountryListMvp.RemoteDataSource
import rx.Observable
import javax.inject.Inject

internal class CountryListInteractor @Inject constructor(
        private val mLocalDataSource: CountryListMvp.LocalDataSource,
        private val mRemoteDataSource: RemoteDataSource
) : CountryListMvp.Interactor {

    override fun getCountryList(): Observable<List<Country>>? = if (mLocalDataSource.isLocalDataPresent()) {
        mLocalDataSource.getCountryList()
    } else {
        mRemoteDataSource.getCountryList()?.also { observable ->
            saveCountry(observable)
        }
    }

    private fun saveCountry(observable: Observable<List<Country>>) {
        observable.map { countries ->
            countries?.map { country ->
                mLocalDataSource.saveCountry(country)
            }
        }
    }
}